types/core.go

```diff
@@ -0,0 +1,335 @@
+package types
+
+import (
+	"fmt"
+
+	"github.com/kong/deck/crud"
+	"github.com/kong/deck/konnect"
+	"github.com/kong/deck/state"
+	"github.com/kong/go-kong/kong"
+)
+
+func eventFromArg(arg crud.Arg) crud.Event {
+	event, ok := arg.(crud.Event)
+	if !ok {
+		panic("unexpected type, expected diff.Event")
+	}
+	return event
+}
+
+type Entity interface {
+	Type() string
+	CRUDActions() crud.Actions
+	PostProcessActions() crud.Actions
+}
+
+type entityImpl struct {
+	typ                string
+	cRUDActions        crud.Actions // needs to set client
+	postProcessActions crud.Actions // needs currentstate Set
+}
+
+func (e entityImpl) Type() string {
+	return e.typ
+}
+
+func (e entityImpl) CRUDActions() crud.Actions {
+	return e.cRUDActions
+}
+
+func (e entityImpl) PostProcessActions() crud.Actions {
+	return e.postProcessActions
+}
+
+type EntityOpts struct {
+	CurrentState  *state.KongState
+	TargetState   *state.KongState
+	KongClient    *kong.Client
+	KonnectClient *konnect.Client
+}
+
+const (
```

Comment:
Entity types feel like a good case for an enumeration type to me. That might be better than an opaque `string`.

Suggestion: `type EntityType string` and define these consts below as `EntityType`.

Reply:
Suggestion accepted. Is it okay if I do it in a follow up refactor after this and #436 PRs merge in? Tracking in https://github.com/Kong/deck/issues/440.